crates/ruma-events/src/direct.rs

```diff
@@ -7,23 +7,80 @@ use std::{
     ops::{Deref, DerefMut},
 };
 
-use ruma_common::{OwnedRoomId, OwnedUserId};
-use ruma_macros::EventContent;
+use ruma_common::{IdParseError, OwnedRoomId, OwnedUserId, UserId};
+use ruma_macros::{EventContent, IdZst};
 use serde::{Deserialize, Serialize};
 
+/// An user identifier, it can be a MXID or a third-party identifier
+/// like an email or a phone number.
+///
+/// There is no validation on this type, any string is allowed,
+/// but you can use `to_user_id` to try to get a MXID.
+#[repr(transparent)]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, IdZst)]
+pub struct DirectUserIdentifier(str);
+
+impl TryFrom<OwnedDirectUserIdentifier> for OwnedUserId {
```

Comment:
We should be able to have `impl<'a> TryFrom<&'a DirectUserIdentifier> for &'a UserId` too.

Maybe to make the conversions more apparent we should also implement:

```rust
impl DirectUserIdentifier {
  pub fn as_user_id(&self) -> Option<&UserId> {}
}

impl OwnedDirectUserIdentifier {
  pub fn as_user_id(&self) -> Option<&UserId> {}
  pub fn into_user_id(self) -> Option<OwnedUserId> {}
}
```